migrations/programs-startDate-EndDate-6.0/setProgramsStartDateAndEndDate.js

```diff
@@ -4,64 +4,147 @@ let MongoClient = require("mongodb").MongoClient;
+  try {
+    console.log("-----------------------Started----------------------");
+    // get all active programs with createdAt value
+    let collectionDocs = await db
+      .collection("programs")
+      .find({
+        status: "active",
+      })
+      .project({ _id: 1, createdAt: 1, components: 1 })
+      .toArray();
 
-                let monthToAdd = 12; // ---------------------set to one year -----------------------------
-                let id = currentChunk[currentChunkIndex]._id;
-                const startDate = currentChunk[currentChunkIndex].createdAt
-                const endDate = new Date(currentChunk[currentChunkIndex].createdAt);
-                endDate.setFullYear( endDate.getFullYear(), endDate.getMonth() + monthToAdd );
+    //reduce array to small chunks
+    let chunkOfprogramsDetails = _.chunk(collectionDocs, 50);
 
-                //update programs
-                await db.collection("programs").findOneAndUpdate(
-                    { '_id': id }, 
-                    { $set: { startDate: startDate, endDate : endDate} }
-                );
-                console.log("program Updated : ",id)
-                updatedProgramIds.push(id)
+    //loop each chunk
+    for (
+      let chunkPointer = 0;
+      chunkPointer < chunkOfprogramsDetails.length;
+      chunkPointer++
+    ) {
+      let currentChunk = chunkOfprogramsDetails[chunkPointer];
+      //loop each program details
+      for (
+        let currentChunkIndex = 0;
+        currentChunkIndex < currentChunk.length;
+        currentChunkIndex++
+      ) {
+        let id = currentChunk[currentChunkIndex]._id;
+        let endDate = "";
+        let startDate = "";
+        let monthToAdd = 12; // ---------------------set to one year -----------------------------
+        //checking if program has components or not
+        if (currentChunk[currentChunkIndex].components.length === 0) {
+          //if it will not have components then createdAt will become startDate and endDate will 1 year from startDate
+          startDate = currentChunk[currentChunkIndex].createdAt;
+          endDate = new Date(currentChunk[currentChunkIndex].createdAt);
+          endDate.setFullYear(
+            endDate.getFullYear(),
+            endDate.getMonth() + monthToAdd
+          );
+        } else {
+          //if program has components then will query solutions for components and will get details of solutions
+          const solutions = await db
+            .collection("solutions")
+            .find({
+              _id: { $in: currentChunk[currentChunkIndex].components },
+            })
+            .project({
+              _id: 1,
+              startDate: 1,
+              endDate: 1,
+              createdAt: 1,
+              status: 1,
+            })
+            .sort({ endDate: -1 })
+            .toArray();
+          //looping throught each solution to check if solution need update or not
+          solutions.forEach(async (solution) => {
```

Comment:
@ankitpws Convert this to simple for loop

Reply:
Updated.